accepted/2021/InboxSourceGenerators.md

```diff
@@ -0,0 +1,126 @@
+
+### Source generator versioning
+
+Source generators should be strongly named and update assembly version with every public stable release (GA and servicing).  This ensures that the publicly shipped source generators can be loaded side-by-side in the same process and garuntees 
+
+### Source generator compatibility concerns
+
+Source generators should only use public API in framework components they extend.  This public API already needs to meet the strict binary compatibility guarantees as framework API.  As such libraries containing source-generated code will be binary-compatible with future releases, just as if the user wrote this code directly.
+
+Source generators should not use public API with "reserved" or "undocumented" functionality in attempts to avoid public API guarantees.  Libraries paired with source generators need to provide the same guarantees as libraries alone do today.
+
+Source generators should avoid emitting public API in user's projects.  Such API can create a compatibility contract in that user's library which needs to be maintained.  If a source generator produces public API in a user's project by design, that public API must be strictly maintained following the same rules as framework components across versions of the source generator.  Care should be taken to the public API which is generated and should be reviewed by the Framework Design Council.  A similar constraint applies to internal API: users will have source that expects to interact with the generated internal API and the user will expect that code to continue to compile between versions.
+
+Generated code should not be generated in System namespaces, it should prefer the user's namespace.  Generated code should be triple-slashed commented and be free from compiler warnings.  Generated code should avoid name colisions with user code and provide affordances for resolving name conflicts. 
+
+### Functionality concerns for source generators
+
+Though the following are not specific to inbox source generators, these concerns are worth reviewing to describe how they apply to framework specific source generators.
+
+Source generators target .NETStandard and must run on both .NETFramework and .NETCore.  
+
+Source generators should limit their dependencies.  There is no scheme for encoding framework-specific nor runtime-specific implementations of dependencies or source generators based on the runtime environment so source generators should avoid any dependencies on assemblies which need to differ by framework or runtime.  Source generators may depend on packages which are known dependencies of the compiler and should depend on versions less than or equal to that provided by the compiler and should not include that dependency in their package.
```

Comment:
> There is no scheme for encoding framework-specific nor runtime-specific implementations of dependencies or source generators based on the runtime environment

This doesn't seem like a particularly difficult problem for the SDK to solve if it needed to, no?

Reply:
It's actually probably pretty hard.  And source generators that are delivered as NuGet packages don't support these types of dependencies, so we shouldn't try to add them to in-box source generators either (unless / until the support for NuGet source generators changes).

Reply:
Right: I'm specifically calling this out as hard and should be avoided.  I wanted to call out that "provided" packages can be used, like System.Reflection.Metadata, System.Collections.Immutable, System.Memory.

For this problem to be "solved" you need the host to solve it, so roslyn would need to have a more complicated plugin model involving something like deps files or a custom ALC, dll.configs, and something brand new that can multiplex between frameworks at runtime.  Let's not go there.